core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -278,17 +303,38 @@ public StreamObserver<UploadSnapshotPRequest> receiveSnapshotFromFollower(
     }
     observer.getFuture()
         .thenApply(termIndex -> {
-          mDownloadedSnapshot = observer.getSnapshotToInstall();
-          transitionState(DownloadState.STREAM_DATA, DownloadState.DOWNLOADED);
-          return termIndex;
+          mRequestDataLock.lock();
+          try {
+            mDownloadedSnapshot = observer.getSnapshotToInstall();
+            transitionState(DownloadState.STREAM_DATA, DownloadState.DOWNLOADED);
+            return termIndex;
+          } finally {
+            // Cancel any pending data requests since the download was successful
+            try {
+              mRequestDataFuture.cancel(true);
+              mRequestDataCondition.signalAll();
+            } finally {
+              mRequestDataLock.unlock();
+            }
+          }
```

Comment:
Why not use `LockResource` instead when locking/unlocking `mRequestDataLock`? I find try-with-resource blocks more legible. Is there a reason why they were not used?
```suggestion
          try (LockResource lr = new LockResource(mRequestDataLock)) {
            mDownloadedSnapshot = observer.getSnapshotToInstall();
            transitionState(DownloadState.STREAM_DATA, DownloadState.DOWNLOADED);
            mRequestDataFuture.cancel(true);
            mRequestDataCondition.signalAll();
            return termIndex;
          }
```